### PR TITLE
feat: Add `oras upload` command (WIP)

### PR DESCRIPTION
**Pull request by TerryHowe**
This is just a POC for download/upload commands for discussion. Related to https://github.com/oras-project/oras/pull/788

### CODE REVIEW

File: cmd/oras/upload.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func runupload(opts uploadOptions, destination string) error {
+	ctx, _ := opts.SetLoggerLevel()
+
+	var uniqueImages []registry.Reference
+	err := filepath.Walk(directory,
```
Comment (qweeah) [inline cmd/oras/upload.go:73]: This looks VERY interesting! The long description is not updated, but looks like this command is trying to upload all the image layout subfolders to a remote registry, using relative path as the repository names?

File: cmd/oras/upload.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			ociPath, layout := strings.CutSuffix(path, "/oci-layout")
+			if layout == true {
+				ociPath, _ = strings.CutPrefix(ociPath, directory+"/")
+				lastIdx := strings.LastIndex(ociPath, "/")
+				reference := registry.Reference{
```
Comment (qweeah) [inline cmd/oras/upload.go:82]: use [Reference](https://pkg.go.dev/oras.land/oras-go/v2@v2.0.0/registry#ParseReference)(ociPath) instead

File: cmd/oras/upload.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		},
+	}
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func runupload(opts uploadOptions, destination string) error {
+	ctx, _ := opts.SetLoggerLevel()
+
+	var uniqueImages []registry.Reference
```
Comment (qweeah) [inline cmd/oras/upload.go:72]: this is an array of repositories?
```suggestion
	var repos []registry. Reference
```

File: cmd/oras/upload.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		to := option.Target{
+			Path:         path.Join(destination, fileReference.Repository),
+			Reference:    fileReference.Reference,
+			RawReference: fileReference.String(),
+			Remote:       opts.To,
+			Type:         option.TargetTypeRemote,
+		}
+
+		to.Reference = ""
+		dst, err := to.NewTarget(opts.Common)
```
Comment (qweeah) [inline cmd/oras/upload.go:138]: Since the target cannot be a oci layout, you can use opts.To.NewRepository directly

File: cmd/oras/upload.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+	var uniqueImages []registry.Reference
+	err := filepath.Walk(directory,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			ociPath, layout := strings.CutSuffix(path, "/oci-layout")
+			if layout == true {
+				ociPath, _ = strings.CutPrefix(ociPath, directory+"/")
```
Comment (qweeah) [inline cmd/oras/upload.go:80]: Looks like you need to 1) create an OCI read-only store 2) get all manifests from the store 3) add manifest reference to `uniqueImages`

File: cmd/oras/upload.go (hunk 1 above)
Comment (TerryHowe) [reply cmd/oras/upload.go:73]: Yes.

**Pull request closed without merging**